File: infumap/src/web/routes/html.rs
// Copyright (C) 2023 The Infumap Authors
// This file is part of Infumap.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

use std::sync::Arc;

use bytes::Bytes;
use http_body_util::combinators::BoxBody;
use hyper::{Request, Response, Method};

use crate::{storage::{db::Db, object}, web::serve::full_body};


pub async fn serve_html_routes(
    db: &Arc<tokio::sync::Mutex<Db>>,
    object_store: &Arc<object::ObjectStore>,
    req: &Request<hyper::body::Incoming>) -> Option<Response<BoxBody<Bytes, hyper::Error>>> {
  match (req.method(), req.uri().path()) {
    (&Method::GET, "/add") => Some(Response::builder().header(hyper::header::CONTENT_TYPE, "text/html").body(full_body(include_str!("../../../dist/add.html"))).unwrap()),
    // (&Method::POST, "/add") => serve_add_post(db, object_store, req),
    _ => None
  }
}

// fn serve_add_post(
//     db: &Arc<tokio::sync::Mutex<Db>>,
//     object_store: &Arc<object::ObjectStore>,
//     req: &Request<hyper::body::Incoming>) -> Option<Response<BoxBody<Bytes, hyper::Error>>> {
  
//   Some(Response::builder().header(hyper::header::CONTENT_TYPE, "text/html").body(full_body(include_str!("../../../dist/upload.html"))).unwrap())
// }

File: web/src/util/html.ts
/*
  Copyright (C) The Infumap Authors
  This file is part of Infumap.

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU Affero General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public License
  along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import { LINE_HEIGHT_PX } from "../constants";


const cache = new Map<String, number>();

export function measureLineCount(s: string, widthBl: number): number {
  const key = s + "-#####-" + widthBl; // TODO (LOW): not foolproof.
  if (cache.has(key)) {
    return cache.get(key)!;
  }
  if (cache.size > 10000) {
    // TODO (LOW): something better than this, though this trivial strategy should be very effective.
    cache.clear();
  }
  const div = document.createElement("div");
  div.setAttribute("style", `line-height: ${LINE_HEIGHT_PX}px; width: ${widthBl*LINE_HEIGHT_PX}px; overflow-wrap: break-word;`);
  const txt = document.createTextNode(s);
  div.appendChild(txt);
  document.body.appendChild(div);
  const lineCount = div.offsetHeight / LINE_HEIGHT_PX;
  document.body.removeChild(div);
  const result = Math.floor(lineCount);
  cache.set(key, result);
  return result;
}
